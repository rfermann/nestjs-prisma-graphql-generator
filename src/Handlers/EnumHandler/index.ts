import type { Enum } from "../../types";
import { NestJSTypes } from "../../types";
import { BaseHandler } from "../BaseHandler";
import { compareObjectValues } from "../BaseHandler/compareFunctions";

export class EnumHandler extends BaseHandler {
  private readonly _enums: Enum[] = [];

  async createBarrelFile(): Promise<void> {
    if (this._enums.length) {
      const sourceFile = this.baseFileGenerator.createSourceFile(`${this.config.paths.enum}/index`);

      sourceFile.addExportDeclarations(
        this._enums.map(({ name }) => {
          return {
            moduleSpecifier: `./${name}`,
            namedExports: [name],
          };
        })
      );
      await sourceFile.save();
    }
  }

  async createFiles(): Promise<void> {
    // eslint-disable-next-line no-restricted-syntax
    for (const { documentation, name: enumName, values } of this._enums) {
      const sourceFile = this.baseFileGenerator.createSourceFile(`${this.config.paths.enum}/${enumName}`);

      this.baseFileGenerator.addNestJSImports({ nestJSImports: [NestJSTypes.RegisterEnumType], sourceFile });

      sourceFile.addEnum({
        isExported: true,
        members: values.map((data) => {
          return {
            name: data.name,
            value: data.value,
          };
        }),
        name: enumName,
      });

      sourceFile.addStatements((writer) =>
        writer
          .writeLine("")
          .writeLine(`${NestJSTypes.RegisterEnumType}(${enumName}, {`)
          .writeLine(`name: "${enumName}",`)
          .conditionalWrite(typeof documentation === "string", `description: "${documentation}"`)
          .writeLine("});")
      );

      sourceFile.formatText();
    }

    await this.baseFileGenerator.save();
  }

  parse(): void {
    this.baseParser.dmmf.datamodel.enums.forEach(({ documentation, name, values }): void => {
      this._enums.push({
        documentation,
        name: this.baseParser.getEnumName(name),
        values: values.map(({ name: value }) => {
          return { name: value, value };
        }),
      });
    });

    this.baseParser.dmmf.schema.enumTypes.prisma.forEach(({ name, values }): void => {
      this._enums.push({
        name: this.baseParser.getEnumName(name),
        values: values.map((value) => {
          return { name: value, value };
        }),
      });
    });

    this._enums.sort((a, b) => compareObjectValues({ a, b, field: "name" }));
  }
}
