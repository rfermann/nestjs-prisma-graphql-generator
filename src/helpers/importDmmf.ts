/* eslint-disable global-require */
/* eslint-disable import/no-dynamic-require */
/* eslint-disable @typescript-eslint/no-var-requires */
/* eslint-disable @typescript-eslint/no-require-imports */
import type { DMMF } from "@prisma/generator-helper";

export const importDmmf = (prismaClientImportPath: string): DMMF.Document =>
  require(prismaClientImportPath).dmmf as DMMF.Document;
