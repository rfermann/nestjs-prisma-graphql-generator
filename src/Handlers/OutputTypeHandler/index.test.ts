// eslint-disable-next-line import/no-nodejs-modules
import { existsSync, readdirSync, readFileSync, rmdirSync } from "fs";

import type { GeneratorConfig } from "../../GeneratorConfig";
import { importDmmf } from "../../helpers";
import { EnumHandler } from "../EnumHandler";

import { OutputTypeHandler } from ".";

describe("OutputTypeHandler", () => {
  // eslint-disable-next-line jest/no-hooks
  beforeAll(async () => {
    if (existsSync(`${process.cwd()}/fixtures/outputTypeHandler/fixtures/`)) {
      rmdirSync(`${process.cwd()}/fixtures/outputTypeHandler/fixtures/`, { recursive: true });
    }
  });
  it("should parse DMMF properly", async () => {
    expect.assertions(1);

    expect(true).toBe(true);
  });

  it("should parse output types and create correct shared files from datamodel", async () => {
    expect.assertions(2);

    const config: GeneratorConfig = {
      basePath: `${process.cwd()}/fixtures/outputTypeHandler/fixtures/output1`,
      includePrismaSelect: false,
      inputArgumentsName: "",
      inputTypeDecoratorsPath: "",
      paths: {
        enums: "enums",
        inputTypeDecorators: "inputTypeDecorators",
        inputTypes: "inputTypes",
        model: "model",
        outputTypes: "outputTypes",
        resolvers: "resolvers",
        shared: "shared",
      },
      prismaClientImportPath: `${process.cwd()}/node_modules/@prisma/client`,
      prismaServiceImport: "",
      prismaServiceImportPath: "",
    };

    const enumHandler = new EnumHandler({
      config,
      dmmf: importDmmf(config.prismaClientImportPath),
    });

    const outputTypeHandler = new OutputTypeHandler({
      config,
      dmmf: importDmmf(config.prismaClientImportPath),
    });

    enumHandler.parse();

    outputTypeHandler.parse(enumHandler.getEnums());
    await outputTypeHandler.createFiles();

    const sharedOutputFiles = readdirSync(`${config.basePath}/${config.paths.shared}/${config.paths.outputTypes}`);

    sharedOutputFiles.forEach((outputType) => {
      const sharedOutputFile = readFileSync(
        `${config.basePath}/${config.paths.shared}/${config.paths.outputTypes}/${outputType}`,
        "utf-8"
      );

      // eslint-disable-next-line jest/prefer-inline-snapshots
      expect(sharedOutputFile).toMatchSnapshot();
    });
    // eslint-disable-next-line jest/prefer-inline-snapshots
    expect(sharedOutputFiles).toMatchSnapshot();
  });

  it("should parse output types and create correct user output files from datamodel", async () => {
    expect.assertions(8);

    const config: GeneratorConfig = {
      basePath: `${process.cwd()}/fixtures/outputTypeHandler/fixtures/output2`,
      includePrismaSelect: false,
      inputArgumentsName: "",
      inputTypeDecoratorsPath: "",
      paths: {
        enums: "enums",
        inputTypeDecorators: "inputTypeDecorators",
        inputTypes: "inputTypes",
        model: "model",
        outputTypes: "outputTypes",
        resolvers: "resolvers",
        shared: "shared",
      },
      prismaClientImportPath: `${process.cwd()}/node_modules/@prisma/client`,
      prismaServiceImport: "",
      prismaServiceImportPath: "",
    };

    const enumHandler = new EnumHandler({
      config,
      dmmf: importDmmf(config.prismaClientImportPath),
    });

    const outputTypeHandler = new OutputTypeHandler({
      config,
      dmmf: importDmmf(config.prismaClientImportPath),
    });

    enumHandler.parse();

    outputTypeHandler.parse(enumHandler.getEnums());
    await outputTypeHandler.createFiles();

    const userOutputFiles = readdirSync(`${config.basePath}/User/${config.paths.outputTypes}`);

    userOutputFiles.forEach((outputType) => {
      const userOutputFile = readFileSync(`${config.basePath}/User/${config.paths.outputTypes}/${outputType}`, "utf-8");

      // eslint-disable-next-line jest/prefer-inline-snapshots
      expect(userOutputFile).toMatchSnapshot();
    });
    // eslint-disable-next-line jest/prefer-inline-snapshots
    expect(userOutputFiles).toMatchSnapshot();
  });

  it("should create correct barrel files for User output types", async () => {
    expect.assertions(2);

    const config: GeneratorConfig = {
      basePath: `${process.cwd()}/fixtures/outputTypeHandler/fixtures/output3`,
      includePrismaSelect: false,
      inputArgumentsName: "",
      inputTypeDecoratorsPath: "",
      paths: {
        enums: "enums",
        inputTypeDecorators: "inputTypeDecorators",
        inputTypes: "inputTypes",
        model: "model",
        outputTypes: "outputTypes",
        resolvers: "resolvers",
        shared: "shared",
      },
      prismaClientImportPath: `${process.cwd()}/node_modules/@prisma/client`,
      prismaServiceImport: "",
      prismaServiceImportPath: "",
    };

    const enumHandler = new EnumHandler({
      config,
      dmmf: importDmmf(config.prismaClientImportPath),
    });

    const outputTypeHandler = new OutputTypeHandler({
      config,
      dmmf: importDmmf(config.prismaClientImportPath),
    });

    enumHandler.parse();

    outputTypeHandler.parse(enumHandler.getEnums());
    await Promise.all([outputTypeHandler.createBarrelFiles(), outputTypeHandler.createFiles()]);

    const userOutputTypeIndexFile = readFileSync(
      `${config.basePath}/User/${config.paths.outputTypes}/index.ts`,
      "utf-8"
    );

    const userOutputFiles = readdirSync(`${config.basePath}/User/${config.paths.outputTypes}`);

    const lines = userOutputTypeIndexFile
      .toString()
      .split("\n")
      .filter((line) => line.length).length;

    expect(lines).toBe(userOutputFiles.length - 1);
    expect(userOutputTypeIndexFile).toMatchInlineSnapshot(`
      "export { AggregateUser } from \\"./AggregateUser\\";
      export { UserAvgAggregateOutputType } from \\"./UserAvgAggregateOutputType\\";
      export { UserCountAggregateOutputType } from \\"./UserCountAggregateOutputType\\";
      export { UserGroupByOutputType } from \\"./UserGroupByOutputType\\";
      export { UserMaxAggregateOutputType } from \\"./UserMaxAggregateOutputType\\";
      export { UserMinAggregateOutputType } from \\"./UserMinAggregateOutputType\\";
      export { UserSumAggregateOutputType } from \\"./UserSumAggregateOutputType\\";
      "
    `);
  });

  it("should create correct barrel files for shared output types", async () => {
    expect.assertions(2);

    const config: GeneratorConfig = {
      basePath: `${process.cwd()}/fixtures/outputTypeHandler/fixtures/output4`,
      includePrismaSelect: false,
      inputArgumentsName: "",
      inputTypeDecoratorsPath: "",
      paths: {
        enums: "enums",
        inputTypeDecorators: "inputTypeDecorators",
        inputTypes: "inputTypes",
        model: "model",
        outputTypes: "outputTypes",
        resolvers: "resolvers",
        shared: "shared",
      },
      prismaClientImportPath: `${process.cwd()}/node_modules/@prisma/client`,
      prismaServiceImport: "",
      prismaServiceImportPath: "",
    };

    const enumHandler = new EnumHandler({
      config,
      dmmf: importDmmf(config.prismaClientImportPath),
    });

    const outputTypeHandler = new OutputTypeHandler({
      config,
      dmmf: importDmmf(config.prismaClientImportPath),
    });

    enumHandler.parse();

    outputTypeHandler.parse(enumHandler.getEnums());
    await Promise.all([outputTypeHandler.createBarrelFiles(), outputTypeHandler.createFiles()]);

    const sharedOutputTypeIndexFile = readFileSync(
      `${config.basePath}/shared/${config.paths.outputTypes}/index.ts`,
      "utf-8"
    );

    const sharedOutputFiles = readdirSync(`${config.basePath}/shared/${config.paths.outputTypes}`);

    const lines = sharedOutputTypeIndexFile
      .toString()
      .split("\n")
      .filter((line) => line.length).length;

    expect(lines).toBe(sharedOutputFiles.length - 1);
    expect(sharedOutputTypeIndexFile).toMatchInlineSnapshot(`
      "export { AffectedRowsOutput } from \\"./AffectedRowsOutput\\";
      "
    `);
  });
});
