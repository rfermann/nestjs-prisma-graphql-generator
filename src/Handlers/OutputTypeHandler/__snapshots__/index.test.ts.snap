// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OutputTypeHandler should parse output types and create correct shared files from datamodel 1`] = `
"import { Field, Int, ObjectType } from \\"@nestjs/graphql\\";

@ObjectType({
  isAbstract: true,
})
export class AffectedRowsOutput {
  @Field(() => Int, {
    nullable: false
  })
  count!: number;
}
"
`;

exports[`OutputTypeHandler should parse output types and create correct shared files from datamodel 2`] = `
Array [
  "AffectedRowsOutput.ts",
]
`;

exports[`OutputTypeHandler should parse output types and create correct user output files from datamodel 1`] = `
"import { Field, ObjectType } from \\"@nestjs/graphql\\";
import { UserAvgAggregateOutputType, UserCountAggregateOutputType, UserMaxAggregateOutputType, UserMinAggregateOutputType, UserSumAggregateOutputType } from \\".\\";

@ObjectType({
  isAbstract: true,
})
export class AggregateUser {
  @Field(() => UserCountAggregateOutputType, {
    nullable: true
  })
  count?: UserCountAggregateOutputType | null;

  @Field(() => UserAvgAggregateOutputType, {
    nullable: true
  })
  avg?: UserAvgAggregateOutputType | null;

  @Field(() => UserSumAggregateOutputType, {
    nullable: true
  })
  sum?: UserSumAggregateOutputType | null;

  @Field(() => UserMinAggregateOutputType, {
    nullable: true
  })
  min?: UserMinAggregateOutputType | null;

  @Field(() => UserMaxAggregateOutputType, {
    nullable: true
  })
  max?: UserMaxAggregateOutputType | null;
}
"
`;

exports[`OutputTypeHandler should parse output types and create correct user output files from datamodel 2`] = `
"import { Field, Float, ObjectType } from \\"@nestjs/graphql\\";

@ObjectType({
  isAbstract: true,
})
export class UserAvgAggregateOutputType {
  @Field(() => Float, {
    nullable: true
  })
  id?: number | null;

  @Field(() => Float, {
    nullable: true
  })
  balance?: number | null;
}
"
`;

exports[`OutputTypeHandler should parse output types and create correct user output files from datamodel 3`] = `
"import { Field, Int, ObjectType } from \\"@nestjs/graphql\\";

@ObjectType({
  isAbstract: true,
})
export class UserCountAggregateOutputType {
  @Field(() => Int, {
    nullable: false
  })
  id!: number;

  @Field(() => Int, {
    nullable: false
  })
  email!: number;

  @Field(() => Int, {
    nullable: false
  })
  name!: number;

  @Field(() => Int, {
    nullable: false
  })
  type!: number;

  @Field(() => Int, {
    nullable: false
  })
  active!: number;

  @Field(() => Int, {
    nullable: false
  })
  balance!: number;

  @Field(() => Int, {
    nullable: false
  })
  rawData!: number;

  @Field(() => Int, {
    nullable: false
  })
  _all!: number;
}
"
`;

exports[`OutputTypeHandler should parse output types and create correct user output files from datamodel 4`] = `
"import { Field, Float, Int, ObjectType } from \\"@nestjs/graphql\\";
import { Prisma } from \\"@prisma/client\\";
import { GraphQLJSON } from \\"graphql-type-json\\";
import { UserTypeEnum } from \\"../../enums\\";
import { UserAvgAggregateOutputType, UserCountAggregateOutputType, UserMaxAggregateOutputType, UserMinAggregateOutputType, UserSumAggregateOutputType } from \\".\\";

@ObjectType({
  isAbstract: true,
})
export class UserGroupByOutputType {
  @Field(() => Int, {
    nullable: false
  })
  id!: number;

  @Field(() => String, {
    nullable: false
  })
  email!: string;

  @Field(() => String, {
    nullable: true
  })
  name?: string | null;

  @Field(() => UserTypeEnum, {
    nullable: false
  })
  type!: \\"USER\\" | \\"ADMIN\\";

  @Field(() => Boolean, {
    nullable: false
  })
  active!: boolean;

  @Field(() => Float, {
    nullable: false
  })
  balance!: number;

  @Field(() => GraphQLJSON, {
    nullable: false
  })
  rawData!: Prisma.JsonValue;

  @Field(() => UserCountAggregateOutputType, {
    nullable: true
  })
  count?: UserCountAggregateOutputType | null;

  @Field(() => UserAvgAggregateOutputType, {
    nullable: true
  })
  avg?: UserAvgAggregateOutputType | null;

  @Field(() => UserSumAggregateOutputType, {
    nullable: true
  })
  sum?: UserSumAggregateOutputType | null;

  @Field(() => UserMinAggregateOutputType, {
    nullable: true
  })
  min?: UserMinAggregateOutputType | null;

  @Field(() => UserMaxAggregateOutputType, {
    nullable: true
  })
  max?: UserMaxAggregateOutputType | null;
}
"
`;

exports[`OutputTypeHandler should parse output types and create correct user output files from datamodel 5`] = `
"import { Field, Float, Int, ObjectType } from \\"@nestjs/graphql\\";
import { UserTypeEnum } from \\"../../enums\\";

@ObjectType({
  isAbstract: true,
})
export class UserMaxAggregateOutputType {
  @Field(() => Int, {
    nullable: true
  })
  id?: number | null;

  @Field(() => String, {
    nullable: true
  })
  email?: string | null;

  @Field(() => String, {
    nullable: true
  })
  name?: string | null;

  @Field(() => UserTypeEnum, {
    nullable: true
  })
  type?: \\"USER\\" | \\"ADMIN\\" | null;

  @Field(() => Boolean, {
    nullable: true
  })
  active?: boolean | null;

  @Field(() => Float, {
    nullable: true
  })
  balance?: number | null;
}
"
`;

exports[`OutputTypeHandler should parse output types and create correct user output files from datamodel 6`] = `
"import { Field, Float, Int, ObjectType } from \\"@nestjs/graphql\\";
import { UserTypeEnum } from \\"../../enums\\";

@ObjectType({
  isAbstract: true,
})
export class UserMinAggregateOutputType {
  @Field(() => Int, {
    nullable: true
  })
  id?: number | null;

  @Field(() => String, {
    nullable: true
  })
  email?: string | null;

  @Field(() => String, {
    nullable: true
  })
  name?: string | null;

  @Field(() => UserTypeEnum, {
    nullable: true
  })
  type?: \\"USER\\" | \\"ADMIN\\" | null;

  @Field(() => Boolean, {
    nullable: true
  })
  active?: boolean | null;

  @Field(() => Float, {
    nullable: true
  })
  balance?: number | null;
}
"
`;

exports[`OutputTypeHandler should parse output types and create correct user output files from datamodel 7`] = `
"import { Field, Float, Int, ObjectType } from \\"@nestjs/graphql\\";

@ObjectType({
  isAbstract: true,
})
export class UserSumAggregateOutputType {
  @Field(() => Int, {
    nullable: true
  })
  id?: number | null;

  @Field(() => Float, {
    nullable: true
  })
  balance?: number | null;
}
"
`;

exports[`OutputTypeHandler should parse output types and create correct user output files from datamodel 8`] = `
Array [
  "AggregateUser.ts",
  "UserAvgAggregateOutputType.ts",
  "UserCountAggregateOutputType.ts",
  "UserGroupByOutputType.ts",
  "UserMaxAggregateOutputType.ts",
  "UserMinAggregateOutputType.ts",
  "UserSumAggregateOutputType.ts",
]
`;
